### PR TITLE
JW-1156: stateful ksync write

### PR DESCRIPTION
**Pull request by sodar**

### CODE REVIEW

File: windows/vr_ksync.c [hunk 1]
```diff
@@ -229,82 +232,120 @@ KsyncHandleWrite(struct ksync_device_context *ctx, uint8_t *buffer, size_t buffe
...
+                userBufferSize -= bytesLeft;
+                writtenBytes += bytesLeft;
+            }
+        } else if (ctx->WrittenBytes >= sizeof(struct nlmsghdr)) {
+            struct nlmsghdr *nlh = (struct nlmsghdr *)ctx->WriteBuffer;
+            ASSERT(nlh->nlmsg_len > ctx->WrittenBytes);
+
+            size_t bytesLeft = nlh->nlmsg_len - ctx->WrittenBytes;
+            if (bytesLeft > ctx->WriteBufferSize - ctx->WrittenBytes) {
+                /* TODO(sodar): How to handle this case properly? */
```
Comment (jablonskim) [inline windows/vr_ksync.c]: Maybe clean up the context here? (Zero `WrittenBytes` etc.)

File: windows/vr_ksync.c [hunk 2]
```diff
@@ -96,51 +106,44 @@ KsyncPopResponse(struct ksync_device_context *ctx)
...
 NTSTATUS
 KsyncDispatchClose(PDEVICE_OBJECT DeviceObject, PIRP Irp)
 {
-    UNREFERENCED_PARAMETER(DeviceObject);
-
-    struct ksync_device_context *ctx = ksync_get_context_from_file_context(Irp);
+    PKSYNC_DEVICE_CONTEXT ctx = KSyncGetContextFromFileContext(Irp);
     ASSERTMSG("KSync device context was not set", ctx != NULL);
 
     ExFreePoolWithTag(ctx, KsyncAllocationTag);
```
Comment (jablonskim) [inline windows/vr_ksync.c]: There may be some responses left in the queue. I think we should free them here. Also, I think that context should be accessed using some lock.

File: windows/vr_ksync.c [hunk 3]
```diff
@@ -229,82 +232,120 @@ KsyncHandleWrite(struct ksync_device_context *ctx, uint8_t *buffer, size_t buffe
...
+        } else if (ctx->WrittenBytes >= sizeof(struct nlmsghdr)) {
+            struct nlmsghdr *nlh = (struct nlmsghdr *)ctx->WriteBuffer;
+            ASSERT(nlh->nlmsg_len > ctx->WrittenBytes);
+
+            size_t bytesLeft = nlh->nlmsg_len - ctx->WrittenBytes;
+            if (bytesLeft > ctx->WriteBufferSize - ctx->WrittenBytes) {
+                /* TODO(sodar): How to handle this case properly? */
+                return KSyncCompleteIrp(Irp, STATUS_UNSUCCESSFUL, 0);
+            }
+            if (userBufferSize < bytesLeft) {
```
Comment (jablonskim) [inline windows/vr_ksync.c]: This also may be simplified a little

File: windows/vr_ksync.c [hunk 4]
```diff
@@ -15,63 +15,66 @@ const WCHAR KsyncDeviceSymLink[] = L"\\DosDevices\\vrouterKsync";
...
 static NDIS_HANDLE    KsyncDeviceHandle   = NULL;
 
-static struct ksync_device_context *
-ksync_alloc_context()
+static PKSYNC_DEVICE_CONTEXT
+KSyncAllocContext()
 {
-    struct ksync_device_context *ctx = ExAllocatePoolWithTag(NonPagedPoolNx,
+    PKSYNC_DEVICE_CONTEXT ctx = ExAllocatePoolWithTag(NonPagedPoolNx,
                                                              sizeof(*ctx),
```
Comment (jablonskim) [inline windows/vr_ksync.c]: Invalid indentation here

File: windows/vr_ksync.c [hunk 5]
```diff
@@ -229,82 +232,120 @@ KsyncHandleWrite(struct ksync_device_context *ctx, uint8_t *buffer, size_t buffe
...
+    if (userBuffer == NULL) {
+        return KSyncCompleteIrp(Irp, STATUS_INSUFFICIENT_RESOURCES, 0);
+    }
 
-            size_t output_buffer_length = io_stack->Parameters.Read.Length;
-            size_t resp_length = resp->message_len;
-            ASSERTMSG("vRouter's response too big", resp_length <= output_buffer_length);
+    PKSYNC_RESPONSE resp = KsyncPopResponse(ctx);
+    if (resp == NULL) {
+        return KSyncCompleteIrp(Irp, STATUS_SUCCESS, 0);
```
Comment (jablonskim) [inline windows/vr_ksync.c]: Is it really expected behavior to return immediately from read method if there is no data to transfer?

File: windows/vr_ksync.c (hunk 2 above)
Comment (jacek-izykowski) [reply windows/vr_ksync.c]: Context access can be improved in separate pull request. This PR doesn't change original behavior.

File: windows/vr_ksync.c [hunk 6]
```diff
@@ -226,85 +242,109 @@ KsyncHandleWrite(struct ksync_device_context *ctx, uint8_t *buffer, size_t buffe
...
+    if (userBuffer == NULL) {
+        return KSyncCompleteIrp(Irp, STATUS_INSUFFICIENT_RESOURCES, 0);
+    }
+
+    ULONG writtenBytes = 0;
+    while (userBufferSize > 0) {
+        if (ctx->WrittenBytes < sizeof(struct nlmsghdr)) {
+            size_t bytesNeeded = sizeof(struct nlmsghdr) - ctx->WrittenBytes;
+            KsyncCopyUserBufferToContext(ctx, bytesNeeded, &userBuffer, &userBufferSize, &writtenBytes);
+        } else if (ctx->WrittenBytes >= sizeof(struct nlmsghdr)) {
```
Comment (jablonskim) [inline windows/vr_ksync.c]: I think that this explicit `if` should be removed

**Pull request merged**